### PR TITLE
[SFN] support for waitForTaskToken and enhancements

### PR DESCRIPTION
**Pull request by MEPalma**
This PR adds support for `waitForTaskToken` task resource types:
- introduces `Callbacks` elements that enable the `sfn` backend to notify awaiting jobs
- adds `StateTaskServiceCallback` to implicitly support `waitForTaskToken` operations for supported services
- adds `test_callback` to test for the baseline `waitForTaskToken` in aws' documentation
- adds `sqs_send_task_success_state_machine` fixture which automatically starts a `sfn` machine which asserts `waitForTaskToken` requests (for testing purposes)
- refactors and cleanups dependencies of state tasks abstractions
- adds support for boto arguments normalisation when these are performed by `sfn` (see `aws-sdk` resolver)
- fixes to `sqs` task resolver to implicitly stringify `MessageBody` inputs (supported by `sfn`)
- adds support for `null` `OutputPath` values
- updates `antlr4` to `4.12.0` from `4.11.1`

### CODE REVIEW

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/resource.py [hunk 1]
```diff
@@ -105,4 +110,16 @@ def __init__(
         )
         self.service_name = service_name
         self.api_name = api_name
-        self.api_action = resource_arn.split(":")[-1]
+
+        arn_parts = resource_arn.split(":")
+        tail_part = arn_parts[-1]
+        tail_parts = tail_part.split(".")
+        self.api_action = tail_parts[0]
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/resource.py:117]: nit: would generally be a good idea in places like this to add comments giving examples (in this case for the `resource_arn`)

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/resource.py [hunk 2]
```diff
@@ -105,4 +110,16 @@ def __init__(
...
+
+        self.condition = None
+        if len(tail_parts) > 1:
+            match tail_parts[-1]:
+                case "waitForTaskToken":
+                    self.condition = ResourceCondition.WaitForTaskToken
+                case unsupported:
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/resource.py:124]:
```suggestion
                case _:
```

File: localstack/services/stepfunctions/provider_v2.py [hunk 3]
```diff
@@ -147,6 +158,46 @@ def describe_state_machine(
...
+        self, context: RequestContext, task_token: TaskToken, output: SensitiveData
+    ) -> SendTaskSuccessOutput:
+        outcome = CallbackOutcomeSuccess(callback_id=task_token, output=output)
+        store = self.get_store(context)
+        for exec in store.executions.values():
+            try:
+                if exec.exec_worker.env.callback_pool_manager.notify(
+                    callback_id=task_token, outcome=outcome
+                ):
+                    return SendTaskSuccessOutput()
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/provider_v2.py:171]: the access pattern here seems a bit noisy/inefficient going through all executions but we can iterate over that at a later point :+1:

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_aws_sdk.py [hunk 4]
```diff
@@ -18,70 +11,43 @@
...
-                    resourceType=self._get_resource_type(),
-                    resource=self.resource.api_action,
-                    output=to_json_str(response),
-                    outputDetails=HistoryEventExecutionDataDetails(truncated=False),
-                )
-            ),
-        )
+        parameter_keys = list(parameters.keys())
+        for parameter_key in parameter_keys:
+            norm_parameter_key = action_normalisers.get(parameter_key, None)
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_aws_sdk.py:46]:
```suggestion
            norm_parameter_key = action_normalisers.get(parameter_key)
```

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_aws_sdk.py [hunk 5]
```diff
@@ -18,70 +11,43 @@
...
-        response = getattr(api_client, api_action)(**parameters) or dict()
-        if response:
-            response.pop("ResponseMetadata", None)
+    def _boto_normalise_parameters(self, api_name: str, api_action: str, parameters: dict) -> None:
+        api_normalisers = self._SFN_TO_BOTO_PARAM_NORMALISERS.get(api_name, None)
+        if not api_normalisers:
+            return
 
-        env.stack.append(response)
+        action_normalisers = api_normalisers.get(api_action, None)
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_aws_sdk.py:40]:
```suggestion
        action_normalisers = api_normalisers.get(api_action)
```

**Review by dominikschubert — APPROVED** [review]
minor nits, otherwise LGTM :rocket:

File: localstack/services/stepfunctions/provider_v2.py (hunk 3 above)
Comment (MEPalma) [reply localstack/services/stepfunctions/provider_v2.py:171]: Yes I definitely agree. I'll not it down and pick it up after this stream of prs. This should make things easier as there is a refactoring of this strategy in one of these (eg: https://github.com/localstack/localstack/pull/8428/files#diff-a5c8a2c872564256e1186b4766cd307a568332bb5f0dd7f709f39f24ad645e4aL16) :)

**Pull request merged**